libraries/Microsoft.Bot.Builder.LanguageGeneration/LGFile.cs

```diff
@@ -0,0 +1,372 @@
+        /// <param name="parameters">new params.</param>
+        /// <param name="templateBody">new template body.</param>
+        /// <returns>new LG resource.</returns>
+        public LGFile UpdateTemplate(string templateName, string newTemplateName, List<string> parameters, string templateBody)
+        {
+            var template = Templates.FirstOrDefault(u => u.Name == templateName);
+            if (template == null)
+            {
+                return this;
+            }
+
+            var templateNameLine = BuildTemplateNameLine(newTemplateName, parameters);
+            var newTemplateBody = ConvertTemplateBody(templateBody);
+            var content = $"{templateNameLine}\r\n{newTemplateBody}\r\n";
+            var startLine = template.ParseTree.Start.Line - 1;
+            var stopLine = template.ParseTree.Stop.Line - 1;
+
+            var newContent = ReplaceRangeContent(Content, startLine, stopLine, content);
+            return LGParser.ParseText(newContent, Id, ImportResolver);
+        }
+
+        /// <summary>
+        /// Add a new template and return LG File.
+        /// </summary>
+        /// <param name="templateName">new template name.</param>
+        /// <param name="parameters">new params.</param>
+        /// <param name="templateBody">new  template body.</param>
+        /// <returns>new lg resource.</returns>
+        public LGFile AddTemplate(string templateName, List<string> parameters, string templateBody)
+        {
+            var template = Templates.FirstOrDefault(u => u.Name == templateName);
+            if (template != null)
+            {
+                throw new Exception($"template {templateName} already exists.");
+            }
+
+            var templateNameLine = BuildTemplateNameLine(templateName, parameters);
+            var newTemplateBody = ConvertTemplateBody(templateBody);
+            var newContent = $"{Content.TrimEnd()}\r\n\r\n{templateNameLine}\r\n{newTemplateBody}\r\n";
+            return LGParser.ParseText(newContent, Id, ImportResolver);
+        }
+
+        /// <summary>
+        /// Delete an exist template.
+        /// </summary>
+        /// <param name="templateName">which template should delete.</param>
+        /// <returns>return the new lg file.</returns>
+        public LGFile DeleteTemplate(string templateName)
+        {
+            var template = Templates.FirstOrDefault(u => u.Name == templateName);
+            if (template == null)
+            {
+                return this;
+            }
+
+            var startLine = template.ParseTree.Start.Line - 1;
+            var stopLine = template.ParseTree.Stop.Line - 1;
+
+            var newContent = ReplaceRangeContent(Content, startLine, stopLine, null);
+            return LGParser.ParseText(newContent, Id, ImportResolver);
+        }
+
+        public override string ToString() => Content;
+
+        public override bool Equals(object obj)
```

Comment:
is there any reason we need this method 'Equals'?

Reply:
`new List<LGFile> { this }.Union(References).SelectMany(x => x.Templates).ToList();`
in this pattern, Linq uses Equals/hashcode to merge the same items.